lib/gitWorkflow.js

```diff
@@ -0,0 +1,187 @@
+        ),
+        readBufferFromFile(this.mergeMsgFile).then(mergeMsg => (this.mergeMsgBuffer = mergeMsg))
+      ])
+      debug('Done backing up merge state!')
+    }
+
+    // Git stash ignores new, untracked files, so back them up separately
+    // Use `git ls-files` to list untracked files:
+    // `--others includes` all untracked files, including ignored files
+    // `--exclude-standard` excludes ignored files, the .git/ dir and so on...
+    let untrackedFiles = await this.execGit(['ls-files', '--others', '--exclude-standard'])
+    if (untrackedFiles) {
+      debug('Detected untracked files:')
+      debug(untrackedFiles)
+      debug('Backing up untracked files...')
+      // Resolve untrackedFiles output into filenames
+      untrackedFiles = untrackedFiles
+        .split('\n')
+        .map(file => normalize(path.resolve(this.cwd, file)))
+      this.untrackedFiles = new Map()
+      const readPromises = []
+      for (const file of untrackedFiles) {
+        // Read all untracked files into buffers, and save them into internal Map
+        readPromises.push(
+          readBufferFromFile(file).then(buffer => this.untrackedFiles.set(file, buffer))
+        )
+      }
+      await Promise.all(readPromises)
+      debug('Done backing up untracked files!')
+    }
+
+    // Get stash of entire original state, including unstaged changes
+    // Keep index so that tasks only work on those files
+    await this.execGit(['stash', 'save', '--quiet', '--include-untracked', '--keep-index', STASH])
+    // Since only staged files are now present, get a diff of unstaged changes
+    // by comparing current index against original stash, but in reverse
+    const original = await this.getBackupStash()
+    this.unstagedDiff = await this.execGit([
+      'diff',
+      '--unified=0',
+      '--no-color',
+      '--no-ext-diff',
+      '-p',
+      original,
+      '-R'
+    ])
+    debug('Done backing up original state!')
+  }
+
+  /**
+   * Resets everything and applies back unstaged and staged changes,
+   * possibly with modifications by tasks
+   *
+   * @param {Object} [options]
+   * @returns {Promise<void>}
+   */
+  async restoreUnstagedChanges() {
+    debug('Restoring unstaged changes...')
+    if (this.unstagedDiff) {
+      try {
+        await this.execGit(gitApplyArgs, { input: `${this.unstagedDiff}\n` })
+      } catch (error) {
+        debug('Error when restoring changes:')
+        debug(error)
+        debug('Retrying with 3-way merge')
+        // Retry with `--3way` if normal apply fails
+        await this.execGit([...gitApplyArgs, '--3way'], { input: `${this.unstagedDiff}\n` })
+      }
+    }
+    debug('Done restoring unstaged changes!')
+
+    if (this.untrackedFiles) {
+      debug('Restoring untracked files...')
+      const writePromises = []
+      this.untrackedFiles.forEach((buffer, file) => {
+        writePromises.push(writeBufferToFile(file, buffer))
+      })
+      await Promise.all(writePromises)
```

Comment:
Could be rewritten with `map` and without `writePromises` variable?

Reply:
Very good, will fix.

Reply:
Actually, `this.untrackedFiles` is a Javascript `Map` and cannot be treated like a array in this case.